### PR TITLE
[flang] Add stack reclaim pass to reclaim allocas in loop

### PR DESCRIPTION
**Pull request by clementval**
Some passes in the flang pipeline are creating `fir.alloca` operation like `hlfir.concat`. When these allocas are located in a loop, the stack can quickly be used too much leading to segfaults. 

This behavior can be seen in https://github.com/jacobwilliams/json-fortran/blob/master/src/tests/jf_test_36.F90

This patch insert a call to LLVM stacksave/stackrestore in the body of the loop to reclaim the alloca in its scope. 

This PR is an alternative implementation to #95173

### CODE REVIEW

**Review by vzakhari — APPROVED** [review]
LGTM.  We should improve this for only constant-sized allocas case in future.

**Comment by clementval** [comment]
> LGTM. We should improve this for only constant-sized allocas case in future.

Yeah I started by only doing it for alloca with dynamic size but in the example mentioned in the description we are generating alloca of descriptor not really dynamic allocas. 

So maybe we should check for dynamic size or descriptor?

**Comment by vzakhari** [comment]
> > LGTM. We should improve this for only constant-sized allocas case in future.
> 
> Yeah I started by only doing it for alloca with dynamic size but in the example mentioned in the description we are generating alloca of descriptor not really dynamic allocas.
> 
> So maybe we should check for dynamic size or descriptor?

Yes, we should probably have a utility `mayBeDynamicAlloca` that will check for the shape/len operands.

**Review by tblah — COMMENTED** [review]
Thank you for running this on all top level operations! One minor comment

**Review by jeanPerier — APPROVED** [review]
Thanks for moving this into a generic pass!
LGTM, please take care of @tblah comment.

File: flang/lib/Optimizer/Transforms/StackReclaim.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+void StackReclaimPass::runOnOperation() {
+  auto *op = getOperation();
+  auto *context = &getContext();
+  mlir::OpBuilder builder(context);
+  mlir::Type voidPtr = mlir::LLVM::LLVMPointerType::get(context);
+
+  op->walk([&](mlir::Operation *op) {
+    if (!mlir::isa<fir::DoLoopOp>(op))
+      return;
```
[the remark this answers is not in the document]
Comment (clementval) [reply flang/lib/Optimizer/Transforms/StackReclaim.cpp]: Yeah! It's cleaner indeed. Thanks

**Review by tblah — APPROVED** [review]
Thanks!

**Pull request merged**